File: go/modcdp/translate/translate.go
// MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
// - ./js/src/translate/translate.ts
// - ./python/modcdp/translate/translate.py
package translate

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/browserbase/modcdp/go/modcdp/types"
)

const UpstreamEventBindingName = "__ModCDP_event_from_upstream__"
const CustomEventBindingName = "__ModCDP_custom_event__"

func DefaultClientRoutes() map[string]string {
	return map[string]string{
		"Mod.*":    "service_worker",
		"Custom.*": "service_worker",
		"*.*":      "service_worker",
	}
}

func RouteFor(method string, routes map[string]string) string {
	if route, ok := routes[method]; ok {
		return route
	}
	bestPrefixLen := -1
	bestRoute := ""
	for pattern, route := range routes {
		if pattern == "*.*" || !strings.HasSuffix(pattern, ".*") {
			continue
		}
		prefix := pattern[:len(pattern)-1]
		if strings.HasPrefix(method, prefix) && len(prefix) > bestPrefixLen {
			bestPrefixLen = len(prefix)
			bestRoute = route
		}
	}
	if bestPrefixLen >= 0 {
		return bestRoute
	}
	if route, ok := routes["*.*"]; ok {
		return route
	}
	return "direct_cdp"
}

func stringValue(value any) string {
	if typed, ok := value.(string); ok {
		return typed
	}
	return ""
}

func callFunctionParams(functionDeclaration string) map[string]any {
	return map[string]any{
		"functionDeclaration": functionDeclaration,
		"awaitPromise":        true,
		"returnByValue":       true,
	}
}

func wrapCustomCommand(method string, params map[string]any, sessionID any) map[string]any {
	p, _ := json.Marshal(params)
	runtimeParams := callFunctionParams(`async function(method, paramsJson, cdpSessionId) { return JSON.stringify(await globalThis.ModCDP.handleCommand(method, JSON.parse(paramsJson), cdpSessionId)); }`)
	runtimeParams["arguments"] = []map[string]any{{"value": method}, {"value": string(p)}, {"value": sessionID}}
	return runtimeParams
}

func wrapServiceWorkerCommand(method string, params map[string]any, sessionID string) []types.TranslatedStep {
	if params == nil {
		params = map[string]any{}
	}
	var cdpSessionID any
	if paramsSessionID, _ := params["cdpSessionId"].(string); paramsSessionID != "" {
		cdpSessionID = paramsSessionID
	} else if sessionID != "" {
		cdpSessionID = sessionID
	}
	return []types.TranslatedStep{{Method: "Runtime.callFunctionOn", Params: wrapCustomCommand(method, params, cdpSessionID), Unwrap: "runtime_json"}}
}

func WrapCommandIfNeeded(method string, params map[string]any, routes map[string]string, sessionID string) (types.TranslatedCommand, error) {
	route := RouteFor(method, routes)
	if route == "direct_cdp" {
		return types.TranslatedCommand{Route: route, Target: "direct_cdp", Steps: []types.TranslatedStep{{Method: method, Params: params, SessionID: sessionID}}}, nil
	}
	if route == "service_worker" {
		return types.TranslatedCommand{Route: route, Target: "service_worker", Steps: wrapServiceWorkerCommand(method, params, sessionID)}, nil
	}
	return types.TranslatedCommand{}, fmt.Errorf("unsupported client route %q for %s", route, method)
}

func UnwrapResponseIfNeeded(result map[string]any, unwrap string) (any, error) {
	if unwrap != "runtime" && unwrap != "runtime_json" {
		return result, nil
	}
	if ex, ok := result["exceptionDetails"].(map[string]any); ok {
		msg := ""
		if e, ok := ex["exception"].(map[string]any); ok {
			if d, ok := e["description"].(string); ok {
				msg = d
			}
		}
		if msg == "" {
			if t, ok := ex["text"].(string); ok {
				msg = t
			}
		}
		if msg == "" {
			msg = "Runtime.evaluate failed"
		}
		return nil, fmt.Errorf("%s", msg)
	}
	inner, _ := result["result"].(map[string]any)
	value := inner["value"]
	if unwrap == "runtime_json" {
		if raw, ok := value.(string); ok {
			var decoded any
			if err := json.Unmarshal([]byte(raw), &decoded); err != nil {
				return nil, err
			}
			return decoded, nil
		}
	}
	return value, nil
}

func UnwrapEventIfNeeded(method string, params map[string]any, sessionID string, ourSessionID string) (*types.UnwrappedModCDPEvent, bool) {
	if method != "Runtime.bindingCalled" {
		return nil, false
	}
	name, _ := params["name"].(string)
	payloadStr, _ := params["payload"].(string)
	var payload map[string]any
	if err := json.Unmarshal([]byte(payloadStr), &payload); err != nil || payload == nil {
		return nil, false
	}
	isUpstreamEventBinding := name == UpstreamEventBindingName
	isCustomEventBinding := name == CustomEventBindingName
	if !isUpstreamEventBinding && !isCustomEventBinding {
		return nil, false
	}
	payloadEvent, _ := payload["event"].(string)
	if payloadEvent == "" {
		return nil, false
	}
	resolvedEvent := payloadEvent
	if resolvedEvent == UpstreamEventBindingName || resolvedEvent == CustomEventBindingName {
		return nil, false
	}
	sourceSessionID := sessionID
	if payloadSessionID, ok := payload["cdpSessionId"].(string); ok {
		sourceSessionID = payloadSessionID
	}
	var sourceSessionIDPtr *string
	if sourceSessionID != "" {
		sourceSessionIDPtr = &sourceSessionID
	}
	if data, ok := payload["data"]; ok {
		return &types.UnwrappedModCDPEvent{Event: resolvedEvent, Data: data, SessionID: sourceSessionIDPtr}, true
	}
	return &types.UnwrappedModCDPEvent{Event: resolvedEvent, Data: payload, SessionID: sourceSessionIDPtr}, true
}

func EncodeBindingPayload(payload types.ModCDPBindingPayload) (string, error) {
	encoded, err := json.Marshal(struct {
		Event        string  `json:"event"`
		Data         any     `json:"data"`
		CDPSessionID *string `json:"cdpSessionId"`
	}{
		Event:        payload.Event,
		Data:         payload.Data,
		CDPSessionID: payload.CDPSessionID,
	})
	if err != nil {
		return "", err
	}
	return string(encoded), nil
}

File: python/modcdp/translate/translate.py
# MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
# Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
# - ./js/src/translate/translate.ts
# - ./go/modcdp/translate/translate.go
"""Pure ModCDP <-> CDP translation helpers for the Python client."""

import json

from ..types.modcdp import (
    ModCDPBindingPayload,
    ModCDPRoutes,
    ProtocolParams,
    ProtocolPayload,
    ProtocolResult,
    RuntimeCallFunctionOnParams,
    TranslatedCommand,
    TranslatedStep,
    UnwrappedModCDPEvent,
    _isObjectMap,
)

UPSTREAM_EVENT_BINDING_NAME = "__ModCDP_event_from_upstream__"
CUSTOM_EVENT_BINDING_NAME = "__ModCDP_custom_event__"

DEFAULT_CLIENT_ROUTES: ModCDPRoutes = {
    "Mod.*": "service_worker",
    "Custom.*": "service_worker",
    "*.*": "service_worker",
}


def route_for(method: str, routes: ModCDPRoutes | None = None) -> str:
    routes = routes or {}
    if method in routes:
        return routes[method]
    best_prefix_len = -1
    best_route = None
    for pattern, route in routes.items():
        if pattern == "*.*" or not pattern.endswith(".*"):
            continue
        prefix = pattern[:-1]
        if method.startswith(prefix) and len(prefix) > best_prefix_len:
            best_prefix_len = len(prefix)
            best_route = route
    if best_route is not None:
        return best_route
    if "*.*" in routes:
        return routes["*.*"]
    return "direct_cdp"


def _optional_string(params: ProtocolParams, name: str) -> str | None:
    value = params.get(name)
    if value is None:
        return None
    if not isinstance(value, str):
        raise TypeError(f"{name} must be a string")
    return value


def _object_or_empty(value: object | None) -> dict[str, object]:
    return value if _isObjectMap(value) else {}


def _call_function_params(function_declaration: str) -> RuntimeCallFunctionOnParams:
    return {
        "functionDeclaration": function_declaration,
        "awaitPromise": True,
        "returnByValue": True,
    }


def _wrap_custom_command(method: str, params: ProtocolParams, session_id: str | None) -> RuntimeCallFunctionOnParams:
    runtime_params = _call_function_params(
        "async function(method, paramsJson, cdpSessionId) { "
        "return JSON.stringify(await globalThis.ModCDP.handleCommand(method, JSON.parse(paramsJson), cdpSessionId)); "
        "}"
    )
    runtime_params["arguments"] = [{"value": method}, {"value": json.dumps(params)}, {"value": session_id}]
    return runtime_params


def _wrap_service_worker_command(
    method: str,
    params: ProtocolParams,
    cdp_session_id: str | None = None,
) -> list[TranslatedStep]:
    return [
        TranslatedStep(
            method="Runtime.callFunctionOn",
            params=_wrap_custom_command(method, params, _optional_string(params, "cdpSessionId") or cdp_session_id),
            unwrap="runtime_json",
        )
    ]


def wrap_command_if_needed(
    method: str,
    params: ProtocolParams | None = None,
    *,
    routes: ModCDPRoutes | None = None,
    cdp_session_id: str | None = None,
) -> TranslatedCommand:
    params = params or {}
    route = route_for(method, routes or DEFAULT_CLIENT_ROUTES)
    if route == "direct_cdp":
        step = TranslatedStep(method=method, params=params)
        if cdp_session_id:
            step.sessionId = cdp_session_id
        return TranslatedCommand(route=route, target="direct_cdp", steps=[step])
    if route == "service_worker":
        return TranslatedCommand(
            route=route,
            target="service_worker",
            steps=_wrap_service_worker_command(method, params, cdp_session_id),
        )
    raise RuntimeError(f"Unsupported client route '{route}' for {method}")


def _unwrap_evaluate_response(result: ProtocolResult) -> object:
    if result.get("exceptionDetails"):
        ex = _object_or_empty(result.get("exceptionDetails"))
        exception = _object_or_empty(ex.get("exception"))
        description = exception.get("description")
        text = ex.get("text")
        msg = (
            description
            if isinstance(description, str)
            else text
            if isinstance(text, str)
            else "Runtime.evaluate failed"
        )
        raise RuntimeError(msg)
    inner = _object_or_empty(result.get("result"))
    return inner.get("value")


def unwrap_response_if_needed(result: ProtocolResult, unwrap: str | None = None) -> object:
    if unwrap == "runtime_json":
        value = _unwrap_evaluate_response(result)
        return json.loads(value) if isinstance(value, str) else value
    return _unwrap_evaluate_response(result) if unwrap == "runtime" else (result or {})


def unwrap_event_if_needed(
    method: str,
    params: ProtocolParams,
    session_id: str | None = None,
    our_session_id: str | None = None,
) -> UnwrappedModCDPEvent | None:
    if method != "Runtime.bindingCalled":
        return None
    binding_name = params.get("name")
    if not isinstance(binding_name, str):
        return None
    raw_payload = params.get("payload")
    if not isinstance(raw_payload, str):
        return None
    try:
        parsed: object = json.loads(raw_payload)
    except json.JSONDecodeError:
        return None
    if not _isObjectMap(parsed):
        return None
    payload: ProtocolPayload = parsed
    is_upstream_event_binding = binding_name == UPSTREAM_EVENT_BINDING_NAME
    is_custom_event_binding = binding_name == CUSTOM_EVENT_BINDING_NAME
    if not is_upstream_event_binding and not is_custom_event_binding:
        return None
    payload_event = payload.get("event")
    if not isinstance(payload_event, str) or not payload_event:
        payload_event = None
    if payload_event is None:
        return None
    resolved_event = payload_event
    if resolved_event == UPSTREAM_EVENT_BINDING_NAME or resolved_event == CUSTOM_EVENT_BINDING_NAME:
        return None
    data = payload["data"] if "data" in payload else payload
    raw_source_session_id = payload.get("cdpSessionId")
    source_session_id = raw_source_session_id if isinstance(raw_source_session_id, str) else session_id
    return UnwrappedModCDPEvent(event=resolved_event, data=data, sessionId=source_session_id)


def encode_binding_payload(payload: ModCDPBindingPayload) -> str:
    return json.dumps(
        {
            "event": payload["event"],
            "data": payload.get("data"),
            "cdpSessionId": payload.get("cdpSessionId"),
        },
        separators=(",", ":"),
    )
